### PR TITLE
Support updating grain data mixture during training

### PR DESCRIPTION
**Pull request by aireenmei**
# Description

See the added "9." in data_input_grain.md about this new feature.

FIXES: b/454051801

# Tests
1. config the new mixture in [grain_mixture.json](https://paste.googleplex.com/5950167197024256)
2. test script [test_grain_mix.sh](https://paste.googleplex.com/4728238901428224), [training log](https://paste.googleplex.com/5055067283652608)
3. Inspect the checkpoints under [gs://aireenmei-multipod/test/grain_ckpt/gemini-test/grain-mixture-test-2025-11-15-06-44-13/checkpoints](https://pantheon.corp.google.com/storage/browser/aireenmei-multipod/test/grain_ckpt/gemini-test/grain-mixture-test-2025-11-15-06-44-13/checkpoints;tab=objects?e=13802955&mods=allow_workbench_image_override&prefix=&forceOnObjectsSortingFiltering=false)

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: docs/guides/data_input_pipeline/data_input_grain.md [hunk 1]
```diff
@@ -76,3 +76,51 @@ In Grain checkpoints, each data-loading host has a corresponding JSON file. For
...
+```
+# In your config or on the command line
+grain_train_files="<path_to_A1>,0.4;<path_to_A2>,0.6"
+```
+
+* **Step 2**: create a mixture configuration file
+
+When you are ready to introduce the new dataset, create a JSON configuration file (e.g., `grain_mixture.json`). This file defines both the original mixture and the new one, along with the desired weights for blending them.
+
+**Important:** The `old_dataset` section in this file must exactly match the sources and weights you used in `grain_train_files` for the initial run.
```
Comment (iindyk) [inline docs/guides/data_input_pipeline/data_input_grain.md:96]: this will not allow removing datasets, right? is that ok?

File: src/MaxText/input_pipeline/_grain_data_processing.py [hunk 2]
```diff
@@ -228,16 +235,47 @@ def make_grain_train_iterator(
...
+          shuffle=config.enable_data_shuffling,
+          shuffle_seed=config.data_shuffle_seed,
+          num_epoch=config.num_epoch,
+          dataloading_host_index=process_indices.index(jax.process_index()),
+          dataloading_host_count=len(process_indices),
+          grain_worker_count=config.grain_worker_count,
+      )
+      old_weight = mixture_config["old_dataset_weight"]
+      new_weight = mixture_config["new_dataset_weight"]
+      train_ds = grain.IterDataset.mix([old_dataset, new_dataset], weights=[old_weight, new_weight])
```
Comment (iindyk) [inline src/MaxText/input_pipeline/_grain_data_processing.py:267]: this will work too, but now that you're using IterDataset.mix, checkpoint will have per-component checkpoint, so I think we could do some surgery there and recover each component separately. That would allow completely changing weights (e.g. changing and removing old weights as well). Is that something you'd be interested in? if so we can discuss further
